src/main/java/httpServer/ListenerAndResponder.java

```diff
@@ -40,46 +43,114 @@ public Socket keepListening() throws IOException {
-
-        String httpRequest;
-        httpRequest = readAllLines(bufferedReader);
+        HTTPRequest httpRequest;
+        httpRequest = readAllLinesAndReturnHttpRequest(bufferedReader);
         if (httpRequest == null) {
             StatusCodes statusCodes = StatusCodes.BAD_REQUEST;
             String responseText = statusCodes.httpResponse;
             printServerResponse(responseText, clientSocket);
         }
         if (!httpRequest.equals("")) {
-            HTTPRequest request = RequestParser.parse(httpRequest);
-            System.out.println(httpRequest);
-            String httpResponse = router.getResponse(request);
-            printServerResponse(httpResponse, clientSocket);
+            HTTPResponse httpResponse = router.getResponse(httpRequest);
+            String responseString = httpResponse.toString();
+            System.err.println(responseString.length());
+            printServerResponse(responseString, clientSocket);
         }
 
     }
 
-    private void printServerResponse(String message, Socket clientSocket) throws IOException {
-            ClientWriteable printer = clientWriterFactory.makePrinter(clientSocket);
-            System.out.println(message);
-            printer.println(message);
-
+    private void printServerResponse(String httpResponse, Socket clientSocket) throws IOException {
+        ClientWriteable printer = clientWriterFactory.makePrinter(clientSocket);
+        System.out.println(httpResponse);
+        printer.print(httpResponse);
+        printer.close();
     }
 
-    private String readAllLines(ClientReadable bufferedReader)  {
-        String line;
-        StringBuilder stringBuilder = new StringBuilder();
+    private HTTPRequest readAllLinesAndReturnHttpRequest(ClientReadable bufferedReader)  {
+
+        String character;
+        RequestBuilder requestBuilder = new RequestBuilder();
+        StringBuilder headers = new StringBuilder();
+        StringBuilder body = new StringBuilder();
+
         try {
-            while ((line = bufferedReader.readLine()) != null)
-            {
-                stringBuilder.append(line);
-                stringBuilder.append(CRLF);
-                if (line.equals("")) { break; }
+            while ((character = bufferedReader.read()) != null) {
+
+                headers.append(character);
+                if (checkIfReachedEndOfHeaders(headers)) {
+                    break;
+                }
             }
-            return stringBuilder.toString();
+
         } catch (IOException e) {
             e.printStackTrace();
-            return stringBuilder.toString();
+            return requestBuilder.build();
+        }
+
+        if (checkIfHeadersIndicateARequestBody(headers)) {
+            int bodyLength = getBodyLength(headers);
+            try {
+                while ((character = bufferedReader.read()) != null) {
+                    if (java.util.Objects.equals(bodyLength, 0)) {
```

Comment:
Did `bodyLength == 0` not work here? A bit weird to be calling a static method on `java.util.Objects`.

Reply:
That works! I'm not sure why I thought it wouldn't